runtime/pendulum/src/xcm_config.rs

```diff
@@ -353,12 +357,90 @@ type Transactor = MultiCurrencyAdapter<
 	DepositToAlternative<PendulumTreasuryAccount, Currencies, CurrencyId, AccountId, Balance>,
 >;
 
+// Workarround for TakeFirstAssetTrader
+use frame_support::traits::tokens::{
+	DepositConsequence, Fortitude, Preservation, Provenance, WithdrawConsequence,
+};
+
+pub struct ConcreteAssets;
+impl fungibles::Mutate<AccountId> for ConcreteAssets {}
+impl fungibles::Balanced<AccountId> for ConcreteAssets {
+	type OnDropCredit = fungibles::DecreaseIssuance<AccountId, Self>;
+	type OnDropDebt = fungibles::IncreaseIssuance<AccountId, Self>;
+}
+impl fungibles::Inspect<AccountId> for ConcreteAssets {
+	type AssetId = <Tokens as fungibles::Inspect<AccountId>>::AssetId;
+	type Balance = <Tokens as fungibles::Inspect<AccountId>>::Balance;
+
+	fn minimum_balance(id: Self::AssetId) -> Self::Balance {
+		Tokens::minimum_balance(id)
+	}
+
+	fn total_issuance(asset_id: Self::AssetId) -> Self::Balance {
+		Tokens::total_issuance(asset_id)
+	}
+
+	fn balance(asset_id: Self::AssetId, account_id: &AccountId) -> Self::Balance {
+		Tokens::balance(asset_id, account_id)
+	}
+
+	fn total_balance(asset_id: Self::AssetId, account_id: &AccountId) -> Self::Balance {
+		Tokens::balance(asset_id, account_id)
+	}
+
+	fn reducible_balance(
+		_: Self::AssetId,
+		_: &AccountId,
+		_: Preservation,
+		_: Fortitude,
+	) -> Self::Balance {
+		todo!()
+	}
+
+	fn can_deposit(
+		_: Self::AssetId,
+		_: &AccountId,
+		_: Self::Balance,
+		_: Provenance,
+	) -> DepositConsequence {
+		todo!()
+	}
+
+	fn can_withdraw(
+		_: Self::AssetId,
+		_: &AccountId,
+		_: Self::Balance,
+	) -> WithdrawConsequence<Self::Balance> {
+		todo!()
+	}
+
+	fn asset_exists(_: Self::AssetId) -> bool {
+		todo!()
+	}
+}
+impl fungibles::Unbalanced<AccountId> for ConcreteAssets {
+	fn handle_dust(_: fungibles::Dust<AccountId, Self>) {
+		todo!()
+	}
+	fn write_balance(
+		_: Self::AssetId,
+		_: &AccountId,
+		_: Self::Balance,
+	) -> Result<Option<Self::Balance>, DispatchError> {
+		todo!()
+	}
+
+	fn set_total_issuance(_: Self::AssetId, _: Self::Balance) {
+		todo!()
+	}
+}
+
```

Comment:
Are we sure that we can leave `todo!()`?

I see [here](https://github.com/paritytech/cumulus/blob/polkadot-v0.9.42/primitives/utility/src/lib.rs#L124) that the trait bound only requires `fungibles::Mutate<AccountId> + fungibles::Balanced<AccountId>,`, is there a need to implement `Unbalanced` and `Inspect` as well?

Reply:
I don't quite remember why this had to be implemented, even though the methods of the traits were not used.
In any case, with the new implementation of [fee per second](https://github.com/pendulum-chain/pendulum/pull/410), `TakeFirstAssetTrader` will not be used anymore, so the workaround is probably not needed anymore.